File: frontend/src/types/project.ts
import type { ProjectStatus, WorkingCalendarType } from "./common";
import type { ProjectMemberCreateRequest } from "./member";
import type { User } from "./user";

/** Краткая карточка проекта для списков и виджетов. */
export interface ProjectSummary {
  /** Уникальный идентификатор проекта. */
  id: string;
  /** Короткий читаемый код проекта. */
  code: string;
  /** Название проекта. */
  name: string;
  /** Текущий статус проекта. */
  status: ProjectStatus;
  /** Текущая фаза реализации проекта (произвольный текст). */
  phase?: string;
  /** Дата начала графика проекта. */
  startDate: string;
  /** Плановый дедлайн (срок сдачи) проекта. */
  deadline: string;
  /** Общий процент выполнения проекта (взвешенный по задачам). */
  progressPercent: number;
  /** Интегральный индекс здоровья проекта (0–100), отображаемый в виде индикатора мониторинга. */
  healthIndex: number;
  /** Количество открытых рисков с критической серьёзностью. */
  criticalRisksCount: number;
}

/** Полная карточка проекта с параметрами календаря и настройками расчёта графика. */
export interface Project extends ProjectSummary {
  /** Описание и бизнес-цель проекта. */
  description?: string;
  /** Заказчик или подразделение, для которого выполняется проект. */
  customerOrg?: string;
  /** Продолжительность проекта в календарных днях (от startDate до deadline). */
  durationCalendarDays: number;
  /** Рабочий календарь, используемый при расчёте сроков задач. */
  workingCalendarType: WorkingCalendarType;
  /** Учитывать ли государственные праздники РФ как нерабочие дни. */
  includePublicHolidays: boolean;
  /**
   * Если true — при сдвиге даты задачи все зависимые задачи по цепочке Finish-to-Start
   * автоматически пересчитываются с сохранением буфера.
   */
  autoRecalculateDependents: boolean;
  /** Если true — задачи с нулевым резервом времени подсвечиваются как критический путь на диаграмме Ганта. */
  highlightCriticalPath: boolean;
  /** Пользователь, создавший проект. */
  createdBy: User;
  /** Дата и время создания проекта. */
  createdAt: string;
  /** Дата и время последнего изменения проекта. */
  updatedAt: string;
}

/** Данные для создания проекта на первом шаге мастера инициации. */
export interface ProjectCreateRequest {
  /** Название проекта. */
  name: string;
  /** Описание и бизнес-цель проекта. */
  description?: string;
  /** Заказчик или подразделение, для которого выполняется проект. */
  customerOrg?: string;
  /** Дата старта графика проекта. */
  startDate: string;
  /** Плановый дедлайн проекта. */
  deadline: string;
  /** Рабочий календарь проекта. */
  workingCalendarType?: WorkingCalendarType;
  /** Учитывать ли государственные праздники РФ как нерабочие дни. */
  includePublicHolidays?: boolean;
  /** Включить автоматический пересчёт зависимых задач при сдвиге дат. */
  autoRecalculateDependents?: boolean;
  /** Подсвечивать критический путь на диаграмме Ганта. */
  highlightCriticalPath?: boolean;
  /** Начальный состав команды проекта, назначаемый на шаге 2 мастера. */
  members?: ProjectMemberCreateRequest[];
  /** Если true — сохранить проект как черновик без перехода к диаграмме Ганта. */
  saveAsDraft?: boolean;
}

/** Набор полей проекта, доступных для частичного обновления (PATCH). */
export interface ProjectUpdateRequest {
  /** Новое название проекта. */
  name?: string;
  /** Новое описание проекта. */
  description?: string;
  /** Новый статус проекта. */
  status?: ProjectStatus;
  /** Новая фаза реализации проекта. */
  phase?: string;
  /** Новая дата старта графика проекта. */
  startDate?: string;
  /** Новый плановый дедлайн проекта. */
  deadline?: string;
  /** Новый рабочий календарь проекта. */
  workingCalendarType?: WorkingCalendarType;
}

/** Параметры фильтрации списка проектов. */
export interface ProjectListParams {
  /** Фильтр по статусу проекта. */
  status?: ProjectStatus;
  /** Подстрока для поиска по названию или коду проекта. */
  search?: string;
}

File: backend/internal/models/project.go
package models

import (
	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Project — проект: контейнер задач, спринтов и вех.
//
// Здесь только то, что задаёт пользователь. Производные показатели
// (progressPercent, healthIndex, durationCalendarDays, criticalRisksCount)
// в таблице не хранятся: они считаются из задач и мгновенно устаревали бы
// после любой правки сроков.
type Project struct {
	Model

	// Code — короткий читаемый код вида TPU-2026-A1B2, генерируется при создании.
	Code        string `gorm:"type:varchar(64);not null;index"`
	Name        string `gorm:"type:varchar(120);not null"`
	Description string `gorm:"type:text"`
	CustomerOrg string `gorm:"type:varchar(255)"`

	Status ProjectStatus `gorm:"type:varchar(20);not null;index"`
	// Phase — текущая фаза реализации, произвольный текст («Фаза 3: Внедрение»).
	Phase string `gorm:"type:varchar(255)"`

	StartDate Date `gorm:"type:date;not null"`
	Deadline  Date `gorm:"type:date;not null"`

	// Настройки расчёта графика — влияют на пересчёт сроков и отрисовку Ганта.
	WorkingCalendarType       WorkingCalendarType `gorm:"type:varchar(10);not null"`
	IncludePublicHolidays     bool                `gorm:"not null;default:true"`
	AutoRecalculateDependents bool                `gorm:"not null;default:true"`
	HighlightCriticalPath     bool                `gorm:"not null;default:true"`

	CreatedByID uuid.UUID `gorm:"type:uuid;not null;index"`
	CreatedBy   *User     `gorm:"foreignKey:CreatedByID"`

	DeletedAt gorm.DeletedAt `gorm:"index"`
}

// TableName фиксирует имя таблицы.
func (Project) TableName() string { return "projects" }

// ProjectMember — участник команды проекта.
//
// ProjectRole и AccessLevel разделены сознательно: роль отвечает на вопрос
// «чем человек занимается», уровень доступа — «что ему разрешено». Аналитик
// может иметь полный доступ, а руководитель со стороны заказчика — только просмотр.
type ProjectMember struct {
	Model

	// Пара (ProjectID, UserID) уникальна: одного человека нельзя добавить
	// в проект дважды. Ограничение стоит в БД, потому что проверка в сервисе
	// не переживает параллельных запросов.
	ProjectID uuid.UUID `gorm:"type:uuid;not null;uniqueIndex:idx_project_members_project_user"`
	UserID    uuid.UUID `gorm:"type:uuid;not null;uniqueIndex:idx_project_members_project_user"`

	Project *Project `gorm:"foreignKey:ProjectID;constraint:OnDelete:CASCADE"`
	User    *User    `gorm:"foreignKey:UserID;constraint:OnDelete:CASCADE"`

	ProjectRole ProjectRole `gorm:"type:varchar(30);not null"`
	// SprintRole — роль в рамках спринтов, свободный текст («Разработчик»).
	SprintRole  string      `gorm:"type:varchar(120)"`
	AccessLevel AccessLevel `gorm:"type:varchar(10);not null"`

	// WeeklyHoursLimit — лимит часов в неделю, база для расчёта загрузки в /workload.
	WeeklyHoursLimit int `gorm:"not null;default:40"`
}

// TableName фиксирует имя таблицы.
func (ProjectMember) TableName() string { return "project_members" }
